### PR TITLE
Oops page and Request access

### PR DESCRIPTION
**Pull request by pabloalba**
Implements https://tree.taiga.io/project/penpot/task/8195

# HOW TO TEST

There are many cases, and the different cases are very similar, but they differ in important nuances. Please read carefully.

## TEST 404

* You will need two users. UserA should be admin of a TeamA with projects and files, and UserB shouldn't be a member of that team.
* You will need two browsers. On the first one log in with UserA. On the second one make sure that you aren't logged in, but log in with UserB when needed.

### Workspace
1.  On the first browser go to a file belonging to TeamA, and copy the URL
2. Create a "bad" URL by changing the last character of the file id (the one just before the '?') for another one.
3. On the second one paste the bad URL
4. You should see the 404 page, with a "Login" button
5. Log in on the second browser with UserB
6. Paste again the bad URL. You should see the 404 page, without the login button, and a "Go to your dashboard" link.
7. Log out on the second browser

### View mode
1. On the first browser go to a file belonging to TeamA, go to its view mode and copy the URL
2. Do the steps 2-7 like in the Workspace

### Dashboard
1.  On the first browser go to a the dashboard of TeamA, and copy the URL
2. Create a "bad" URL by changing the last character of the team id (the one just before the '/projects') for another one.
3. Do the steps 3-7 like in the Workspace

### Workspace (not Your Penpot)
1.  On the first browser go to a file belonging to TeamA, and copy the URL
2. On the second one paste the URL
3. You should see a "workspace-like" background, and a Login popup
4. Log in with userB
5. You should be a popup indicating that you have no access to the file, but can request it. Press the button to request access.
6. On the first browser check in MailCatcher that UserA has received an email with the options of add UserB to the team, or create a view only link.
7. On the first browser press the button on the email to add the user to the team. You should go to the TeamA members page, with the invite window opened, and the email of userB added. DO NOT INVITE THE USER YET!!
8. On the first browser press the button on the email to create a view only link. You should go to the view mode of the file, with the share window opened.
9. Log out on the second browser

### View Mode (not Your Penpot)
1. On the first browser go to a file belonging to TeamA, go to its view mode and copy the URL
2. Do the steps 2-9 like in the Workspace. Check that the background is like the "view-mode".

### Dashboard  (not Your Penpot)
1.  On the first browser go to a the dashboard of TeamA, and copy the URL
2. On the second one paste the URL
3. You should see a "dashboard-like" background, and a Login popup
4. Log in with userB
5. You should be a popup indicating that you have no access to the project, but can request it. Press the button to request access.
6. On the first browser check in MailCatcher that UserA has received an email with the possibility of add UserB to the team.
7. Log out on the second browser

### Workspace (Your Penpot)
1.  On the first browser go to a file belonging to "Your Penpot", and copy the URL
2. On the second one paste the URL
3. You should see a "workspace-like" background, and a Login popup
4. Log in with userB
5. You should be a popup indicating that you have no access to the file, but can request it. Press the button to request access.
6. On the first browser check in MailCatcher that UserA has received an email with the options of move the file to another team, or create a view only link.
9. Log out on the second browser

### View Mode (Your Penpot)
1. On the first browser go to a file belonging to  "Your Penpot", go to its view mode and copy the URL
2. Do the steps 2-5 like in the Workspace. Check that the background is like the "view-mode".
3. On the first browser check in MailCatcher that UserA has received an email with only the option of create a view only link.
9. Log out on the second browser

### Dashboard  (Your Penpot)
1.  On the first browser go to a the dashboard of  "Your Penpot", and copy the URL
2. On the second one paste the URL
3. You should see a "dashboard-like" background, and a Login popup
4. Log in with userB
5. You should be a popup indicating that you have no access to the project, and no option to request it.
7. Log out on the second browser

## TEST AUTO-JOIN
1. On the first browser find one of the emails with the option of add the user to the team. Press the button on the email to add the user to the team. You should go to the TeamA members page, with the invite window opened, and the email of userB added. Invite the user.
2. Check that the user is now a member of the team (the user has auto-joined, without an invitation email)
3. On Mailcatcher, check that UserB has received a mail telling that now is a member of TeamA

### CODE REVIEW

File: backend/src/app/rpc/commands/teams.clj [hunk 1]
```diff
@@ -815,6 +851,58 @@
 
         itoken))))
 
+(defn- add-user-to-team
+  [conn profile team email role]
+
+  (let [team-id (:id team)
+        member (db/get* conn :profile
+                        {:email (str/lower email)}
+                        {:columns [:id :email]})
```
Comment (niwinz) [inline backend/src/app/rpc/commands/teams.clj]: ::sql/columns

File: backend/src/app/rpc/commands/teams.clj [hunk 2]
```diff
@@ -815,6 +851,58 @@
...
+                :invited-by (:fullname profile)
+                :team (:name team)
+                :team-id (:id team)})))
+
+
+(def sql:valid-requests-email
+  "select p.email from team_request tr, profile p
+     where tr.team_id = ?
+       and tr.auto_join_until > now()
+       and tr.requester_id = p.id")
```
Comment (niwinz) [inline backend/src/app/rpc/commands/teams.clj]: 1. use proper and explicit join
2. format sql in this way:
image
all new sql should be formatted using this style
3. Use proper sql aliases with `AS`

File: backend/src/app/rpc/commands/teams.clj [hunk 3]
```diff
@@ -847,13 +935,16 @@
...
-      (when-not (eml/allow-send-emails? conn profile)
-        (ex/raise :type :validation
-                  :code :profile-is-muted
-                  :hint "looks like the profile has reported repeatedly as spam or has permanent bounces"))
+      ;; Check if the current profile is allowed to send emails.
+      (check-valid-email-muted conn profile true)
+
 
-      (let [cfg         (assoc cfg ::db/conn conn)
+      (let [requested      (->> (get-valid-requests-email conn team-id)
```
Comment (niwinz) [inline backend/src/app/rpc/commands/teams.clj]: should be `(into #{} (map :email) (get-valid-requests-email conn team-id))`

File: backend/src/app/rpc/commands/teams.clj [hunk 4]
```diff
@@ -869,6 +960,10 @@
                                           (assoc :role role))))
                                (keep (partial create-invitation cfg)))
                               emails)]
+          ;; For requested invitations, do not send invitation emails, add the user directly to the team
+        (doseq [e emails-to-add]
```
Comment (niwinz) [inline backend/src/app/rpc/commands/teams.clj]: `e` -> `email`

File: backend/src/app/rpc/commands/teams.clj [hunk 5]
```diff
@@ -1007,3 +1102,126 @@
...
+          update set valid_until = ?, auto_join_until = ?, updated_at = now()
+   returning *")
+
+
+(def sql:team-request
+  "select id, (valid_until < now()) as expired
+   from team_request where team_id = ? and requester_id = ?")
+
+(def sql:team-owner
+  "select profile_id
```
Comment (niwinz) [inline backend/src/app/rpc/commands/teams.clj]: The same here, new SQL should be formatted using the new style (caps)

File: backend/src/app/rpc/commands/teams.clj [hunk 6]
```diff
@@ -1007,3 +1102,126 @@
...
+   from team_profile_rel where team_id = ? and is_owner = 't'")
```
Comment (niwinz) [inline backend/src/app/rpc/commands/teams.clj]: is_owner = true

File: backend/src/app/rpc/commands/teams.clj [hunk 7]
```diff
@@ -1007,3 +1102,126 @@
...
+
+
+(defn- create-team-request
+  [{:keys [::db/conn] :as cfg} {:keys [team requester team-owner file is-viewer] :as params}]
+  (let [old-request (->> (db/exec! conn [sql:team-request (:id team) (:id requester)])
```
Comment (niwinz) [inline backend/src/app/rpc/commands/teams.clj]: db/exec-one!

File: backend/src/app/rpc/commands/teams.clj [hunk 8]
```diff
@@ -1007,3 +1102,126 @@
...
+          (or (contains? params :file-id)
+              (contains? params :team-id)))]])
+
+
+(sv/defmethod ::create-team-request
+  "A rpc call that allow to request for an invitations to join the team."
+  {::doc/added "2.2.0"
+   ::sm/params schema:create-team-request}
+  [{:keys [::db/pool] :as cfg} {:keys [::rpc/profile-id file-id team-id is-viewer] :as params}]
+  (db/with-atomic [conn pool]
```
Comment (niwinz) [inline backend/src/app/rpc/commands/teams.clj]: Consider not using with-atomic on new code, db/tx-run should be used

File: frontend/src/app/main/data/common.cljs [hunk 9]
```diff
@@ -155,3 +155,41 @@
...
+    ptk/WatchEvent
+    (watch [_ _ _]
+      (->> (rp/cmd! :get-file-info params)
+           (rx/map on-info)))))
+
+;;;;;;;;;;;;;;;;;;;;;;
+;; Team info
+;;;;;;;;;;;;;;;;;;;;;;
+
+(defn get-team-info
```
Comment (niwinz) [inline frontend/src/app/main/data/common.cljs]: You don't need to define events if you just want to make a request and call the callback, perform the request directlly

File: frontend/src/app/main/errors.cljs [hunk 10]
```diff
@@ -96,16 +95,23 @@
...
 (defmethod ptk/handle-error :authentication
-  [_]
-  (let [msg (tr "errors.auth.unable-to-login")
-        uri (. (. js/document -location) -href)]
-    (st/emit! (du/logout {:capture-redirect true}))
-    (ts/schedule 500 #(st/emit! (msg/warn msg)))
-    (ts/schedule 1000 #(swap! storage assoc :redirect-url uri))))
+  [e]
+  (let [msg        (tr "errors.auth.unable-to-login")
+        uri        (. (. js/document -location) -href)
```
Comment (niwinz) [inline frontend/src/app/main/errors.cljs]: don't use js/document, prefer use app.util.global ns for that purpose

File: frontend/src/app/main/ui/auth/login.cljs [hunk 11]
```diff
@@ -138,9 +138,9 @@
                            :on-success on-success})]
              (st/emit! (du/login-with-ldap params)))))
 
-        on-recovery-request
-        (mf/use-fn
-         #(st/emit! (rt/nav :auth-recovery-request)))]
+        on-recovery-request (or on-recovery-request
+                                (mf/use-fn
```
Comment (niwinz) [inline frontend/src/app/main/ui/auth/login.cljs]: You can't put react-hooks on conditionals, they should be defined statically.

File: backend/src/app/rpc/commands/teams.clj (hunk 1 above)
Comment (pabloalba) [reply backend/src/app/rpc/commands/teams.clj]: fixed

File: backend/src/app/rpc/commands/teams.clj (hunk 4 above)
Comment (pabloalba) [reply backend/src/app/rpc/commands/teams.clj]: fixed

File: backend/src/app/rpc/commands/teams.clj (hunk 6 above)
Comment (pabloalba) [reply backend/src/app/rpc/commands/teams.clj]: fixed

File: backend/src/app/rpc/commands/teams.clj (hunk 7 above)
Comment (pabloalba) [reply backend/src/app/rpc/commands/teams.clj]: fixed

File: backend/src/app/rpc/commands/teams.clj (hunk 2 above)
Comment (pabloalba) [reply backend/src/app/rpc/commands/teams.clj]: fixed

File: backend/src/app/rpc/commands/teams.clj (hunk 5 above)
Comment (pabloalba) [reply backend/src/app/rpc/commands/teams.clj]: fixed

File: frontend/src/app/main/data/common.cljs (hunk 9 above)
Comment (pabloalba) [reply frontend/src/app/main/data/common.cljs:174]: fixed

File: frontend/src/app/main/errors.cljs (hunk 10 above)
Comment (pabloalba) [reply frontend/src/app/main/errors.cljs]: fixed

File: frontend/src/app/main/ui/auth/login.cljs (hunk 11 above)
Comment (pabloalba) [reply frontend/src/app/main/ui/auth/login.cljs]: fixed

File: backend/src/app/rpc/commands/teams.clj (hunk 3 above)
Comment (pabloalba) [reply backend/src/app/rpc/commands/teams.clj]: fixed

File: backend/src/app/rpc/commands/teams.clj [hunk 12]
```diff
@@ -80,6 +82,37 @@
   (cond-> row
     (some? features) (assoc :features (db/decode-pgarray features #{}))))
 
+
+
+(defn- check-valid-email-muted
+  "Check if the member's email is part of the global bounce report."
+  [conn member show?]
```
Comment (niwinz) [inline backend/src/app/rpc/commands/teams.clj]: This function receives the show? parameter, but looks like it is always `true` so we can simplify it.

File: backend/src/app/rpc/commands/teams.clj (hunk 2 above)
Comment (niwinz) [reply backend/src/app/rpc/commands/teams.clj]: It does not looks like fixed.

File: backend/test/backend_tests/rpc_team_test.clj [hunk 13]
```diff
@@ -467,3 +467,146 @@
...
+    (let [owner       (th/create-profile* 1 {:is-active true :email "owner@bar.com"})
+          requester   (th/create-profile* 2 {:is-active true :email "requester@bar.com"})
+          team        (th/create-team* 1 {:profile-id (:id owner)})
+          proj        (th/create-project* 1 {:profile-id (:id owner)
+                                             :team-id (:id team)})
+          file        (th/create-file* 1 {:profile-id (:id owner)
+                                          :project-id (:id proj)})
+
+          pool        (:app.db/pool th/*system*)
+          data        {::th/type :create-team-request
```
Comment (niwinz) [inline backend/test/backend_tests/rpc_team_test.clj]: create-team-request, looks like a not very clear name, can it be `create-team-access-request` ?

File: frontend/src/app/main/ui/auth/recovery_request.cljs [hunk 14]
```diff
@@ -102,3 +102,15 @@
                    :class (stl/css :go-back-link)
                    :data-testid "go-back-link"}
        (tr "labels.go-back")]]]))
+
+
+(mf/defc recovery-sent-page
+  [{:keys [email]}]
```
Comment (niwinz) [inline frontend/src/app/main/ui/auth/recovery_request.cljs:109]: missing `{::mf/props :obj}` metadata, it shuld be present on all new components.

File: frontend/src/app/main/ui/components/forms.cljs [hunk 15]
```diff
@@ -528,6 +528,12 @@
             values (filterv #(:valid %) values)]
         (update-form! values)))
 
+    (mf/with-effect
```
Comment (niwinz) [inline frontend/src/app/main/ui/components/forms.cljs]: Missing the empty `[]` as deps for style consistency.

Comment (niwinz) [reply frontend/src/app/main/ui/components/forms.cljs]: and `invite-mail` probably should be on deps

File: frontend/src/app/main/ui/dashboard/team.cljs [hunk 16]
```diff
@@ -79,7 +79,12 @@
          (fn []
            (st/emit! (modal/show {:type :invite-members
                                   :team team
-                                  :origin :team}))))]
+                                  :origin :team
+                                  :invite-email invite-email}))))]
+
+    (mf/with-effect
```
Comment (niwinz) [inline frontend/src/app/main/ui/dashboard/team.cljs]: The same here, missing deps

File: frontend/src/app/main/ui/static.cljs [hunk 17]
```diff
@@ -10,49 +10,273 @@
...
+
+
+(mf/defc login-dialog
+  [{:keys [show-dialog]}]
+  (let [current-section  (mf/use-state :login)
+        user-email (mf/use-state "")
+        register-token (mf/use-state "")
+        on-close
+        (mf/use-fn
+         #(set! (.-href globals/location) "/"))
```
Comment (niwinz) [inline frontend/src/app/main/ui/static.cljs]: This function is used in many times, I guess it is better to make it a standalone funcion on app.util.router ?

File: frontend/src/app/main/ui/static.cljs [hunk 18]
```diff
@@ -10,49 +10,273 @@
...
+        show-dialog (mf/use-state true)
+        on-success
+        (mf/use-fn
+         #(reset! requested* {:sent true :already-requested false}))
+        on-error
+        (mf/use-fn
+         #(reset! requested* {:sent true :already-requested true}))
+        on-request-access
+        (mf/use-fn
+         (mf/deps file-id)
```
Comment (niwinz) [inline frontend/src/app/main/ui/static.cljs]: missing many deps here

File: frontend/src/app/main/ui/static.cljs [hunk 19]
```diff
@@ -10,49 +10,273 @@
...
+         (fn []
+           (let [params (if (str/empty? file-id) {:team-id team-id} {:file-id file-id :is-viewer (not workspace?)})
+                 mdata  {:on-success on-success :on-error on-error}]
+             (st/emit! (dc/create-team-request (with-meta params mdata))))))]
+
+
+    [:*
+     (if (not (str/empty? file-id))
```
Comment (niwinz) [inline frontend/src/app/main/ui/static.cljs]: file-id is a string? looks strange

File: frontend/src/app/main/ui/static.cljs [hunk 20]
```diff
@@ -10,49 +10,273 @@
...
+      (for [txt content]
+        [:div txt])
+      [:div {:class (stl/css :sign-info)}
+       (when cancel-text
+         [:button {:class (stl/css :cancel-button) :on-click on-close} cancel-text])
+       [:button {:on-click on-click} button-text]]]]))
+
+
+(mf/defc request-access
+  [{:keys [file-id team-id is-default workspace?] :as props}]
```
Comment (niwinz) [inline frontend/src/app/main/ui/static.cljs]: Missing `{::mf/props :obj}` and `:as props` is unused

File: frontend/src/app/main/ui/viewer/header.cljs [hunk 21]
```diff
@@ -159,6 +159,12 @@
         handle-zoom-fit
         (mf/use-fn
          #(st/emit! dv/zoom-to-fit))]
+    (mf/with-effect
```
Comment (niwinz) [inline frontend/src/app/main/ui/viewer/header.cljs]: missing deps

File: frontend/translations/en.po [hunk 22]
```diff
@@ -5322,3 +5328,78 @@ msgstr "Read and modify your libraries and assets."
...
+msgstr "Go to your Penpot"
+
+msgid "not-found.no-permission.project-name"
+msgstr "PROJECT"
+
+msgid "not-found.no-permission.penpot-file"
+msgstr "Penpot file"
+
+msgid "not-found.login.login"
+msgstr "Login"
```
Comment (niwinz) [inline frontend/translations/en.po]: Looks like a good candidate for use `labels.login`

File: frontend/translations/en.po [hunk 23]
```diff
@@ -5322,3 +5328,78 @@ msgstr "Read and modify your libraries and assets."
...
 "Discover [more plugins](%s)"
+
+msgid "not-found.desc-message.error"
+msgstr "404 error"
+
+msgid "not-found.desc-message.doesnt-exist"
+msgstr "This page doesn't exist"
+
+msgid "not-found.copyright"
+msgstr "Kaleidos @2024"
```
Comment (niwinz) [inline frontend/translations/en.po]: Looks like a good candidate for `labels.copyright`

File: backend/src/app/rpc/commands/teams.clj (hunk 12 above)
Comment (pabloalba) [reply backend/src/app/rpc/commands/teams.clj]: fixed

File: frontend/src/app/main/ui/auth/recovery_request.cljs (hunk 14 above)
Comment (pabloalba) [reply frontend/src/app/main/ui/auth/recovery_request.cljs:109]: fixed

File: frontend/src/app/main/ui/components/forms.cljs (hunk 15 above)
Comment (pabloalba) [reply frontend/src/app/main/ui/components/forms.cljs]: fixed

File: frontend/src/app/main/ui/static.cljs (hunk 19 above)
Comment (pabloalba) [reply frontend/src/app/main/ui/static.cljs]: fixed

File: backend/src/app/rpc/commands/teams.clj (hunk 2 above)
Comment (pabloalba) [reply backend/src/app/rpc/commands/teams.clj]: Check again now, please

File: frontend/src/app/main/ui/components/forms.cljs [hunk 24]
```diff
@@ -528,6 +528,13 @@
...
+      (mf/deps [])
```
Comment (niwinz) [inline frontend/src/app/main/ui/components/forms.cljs]: with effect is a macro, the deps are set directy las vector

```
(mf/with-effect [dep dep]
   #_other code
   )
```

**Pull request merged**